recipe/thirdparty/providers/apple.go

```diff
@@ -20,123 +20,93 @@ import (
-	return tpmodels.TypeProvider{
-		ID: appleID,
-		Get: func(redirectURI, authCodeFromRequest *string, userContext supertokens.UserContext) tpmodels.TypeProviderGetResponse {
-			accessTokenAPIURL := "https://appleid.apple.com/auth/token"
-			clientSecret, err := getClientSecret(config.ClientID, config.ClientSecret.KeyId, config.ClientSecret.TeamId, config.ClientSecret.PrivateKey)
+func Apple(input tpmodels.ProviderInput) tpmodels.TypeProvider {
+	if input.ThirdPartyID == "" {
+		input.ThirdPartyID = appleID
+	}
+
+	if input.Config.OIDCDiscoveryEndpoint == "" {
+		input.Config.OIDCDiscoveryEndpoint = "https://appleid.apple.com/"
+	}
+
+	if input.Config.UserInfoMap.FromIdTokenPayload.UserId == "" {
+		input.Config.UserInfoMap.FromIdTokenPayload.UserId = "sub"
+	}
+	if input.Config.UserInfoMap.FromIdTokenPayload.Email == "" {
+		input.Config.UserInfoMap.FromIdTokenPayload.Email = "email"
+	}
+
+	if input.Config.UserInfoMap.FromIdTokenPayload.EmailVerified == "" {
+		input.Config.UserInfoMap.FromIdTokenPayload.EmailVerified = "email_verified"
+	}
+
+	if input.Config.AuthorizationEndpointQueryParams == nil {
+		input.Config.AuthorizationEndpointQueryParams = map[string]interface{}{}
+	}
+
+	if input.Config.AuthorizationEndpointQueryParams["response_type"] == nil {
+		input.Config.AuthorizationEndpointQueryParams["response_type"] = "code"
+	}
+	if input.Config.AuthorizationEndpointQueryParams["response_mode"] == nil {
+		input.Config.AuthorizationEndpointQueryParams["response_mode"] = "form_post"
+	}
+
+	oOverride := input.Override
+
+	input.Override = func(provider *tpmodels.TypeProvider) *tpmodels.TypeProvider {
+		oGetConfig := provider.GetConfig
+		provider.GetConfig = func(clientType *string, input tpmodels.ProviderConfig, userContext supertokens.UserContext) (tpmodels.ProviderConfigForClient, error) {
+			config, err := oGetConfig(clientType, input, userContext)
 			if err != nil {
-				panic(err)
-			}
-			accessTokenAPIParams := map[string]string{
-				"client_id":     config.ClientID,
-				"client_secret": clientSecret,
-				"grant_type":    "authorization_code",
-			}
-			if authCodeFromRequest != nil {
-				accessTokenAPIParams["code"] = *authCodeFromRequest
-			}
-			if redirectURI != nil {
-				accessTokenAPIParams["redirect_uri"] = *redirectURI
+				return tpmodels.ProviderConfigForClient{}, err
 			}
 
-			authorisationRedirectURL := "https://appleid.apple.com/auth/authorize"
-			scopes := []string{"email"}
-			if config.Scope != nil {
-				scopes = config.Scope
+			if len(config.Scope) == 0 {
+				config.Scope = []string{"openid", "email"}
 			}
 
-			var additionalParams map[string]interface{} = nil
-			if config.AuthorisationRedirect != nil && config.AuthorisationRedirect.Params != nil {
-				additionalParams = config.AuthorisationRedirect.Params
+			if config.ClientSecret == "" {
+				clientSecret, err := getClientSecret(config.ClientID, config.AdditionalConfig["clientSecret"].(map[string]interface{}))
```

Comment:
Keep keyId, teamId and privateKey within AdditionalConfig itself and don't further scope it to clientSecret

Reply:
Done